src/algorithms.rs

```diff
@@ -0,0 +1,223 @@
+use std::collections::HashMap;
+
+use substreams::log;
+use substreams_ethereum::pb::eth::v2::{Call, StorageChange, TransactionTrace};
+use substreams::scalar::BigInt;
+use substreams::Hex;
+use crate::abi::erc20::events::Transfer;
+use hex_literal::hex;
+
+const _NULL_ADDRESS: [u8; 20] = hex!("0000000000000000000000000000000000000000");
+const ZERO_STORAGE_PREFIX: [u8; 16] = hex!("00000000000000000000000000000000");
+
+// algorithm #1 (normal case)
+pub fn find_erc20_balance_changes_algorithm1(
+    call: &Call,
+    transfer: &Transfer,
+    keccak_address_map: &StorageKeyToAddressMap,
+) -> Vec<(Vec<u8>, StorageChange)> {
+    let mut out = Vec::new();
+
+    for storage_change in &call.storage_changes {
+        // Check if the transfer matches the storage change balance changes
+        if !is_erc20_valid_balance(transfer, storage_change) {
+            continue;
+        }
+
+        // extract the owner address
+        let owner = match get_owner_from_keccak_address_map(keccak_address_map, &storage_change) {
+            Some(address) => address,
+            None => continue
+        };
+
+        // make sure owner is either the sender or receiver
+        if !is_erc20_valid_address(&owner, transfer) {
+            log::info!("owner={} does not match transfer from={} to={}", Hex(owner), Hex(&transfer.from), Hex(&transfer.to));
+            continue;
+        }
+        out.push((owner, storage_change.clone()));
+    }
+    out
+}
+
+// algorithm #2 (case where storage changes are not in the same call as the transfer event)
+pub fn find_erc20_balance_changes_algorithm2(
+    child_calls: Vec<Call>,
+    transfer: &Transfer,
+    keccak_address_map: &StorageKeyToAddressMap,
+) -> Vec<(Vec<u8>, StorageChange)> {
+    let mut out = Vec::new();
+
+    //get all storage changes for these calls:
+    let mut storage_changes = Vec::new();
+    for call in child_calls.iter() {
+        storage_changes.extend(call.storage_changes.clone());
+    }
```

Comment:
There is so much `.clone()`. Most were there event before. 

I'll try to push on top of branch to remove most of them.

Reply:
https://github.com/pinax-network/substreams-erc20-balance-changes/pull/2/files

Reply:
Thanks for contributing, merged those 👍